UI: Build replay file format so it matches recordings

NikhilNarayana opened the pull request:
### Description
Fixes #2416 
Build replay filename formatting so that it mimics regular recordings and applies the prefix and suffix to only the filename.

### Motivation and Context
Users expect recordings and replays to follow the same filename formatting, but instead replays silently ignore directories by swapping `\\` or `/` with `_`.

### How Has This Been Tested?
The following was done in both simple and advanced output modes. OS is Windows 10 2004.

Set `Replay Buffer Filename Prefix` to `Prefix` and `Replay Buffer Filename Suffix` to `Suffix`.

Left `Filename Formatting` to its default of `%CCYY-%MM-%DD %hh-%mm-%ss` and tested that the output file used `Prefix  %CCYY-%MM-%DD %hh-%mm-%ss Suffix`.

Set `Filename Formatting` to `%CCYY-%MM-%DD/%hh-%mm-%ss` and tested that the output file used `%CCYY-%MM-%DD/Prefix %hh-%mm-%ss Suffix`.

 Set `Filename Formatting` to `%CCYY-%MM-%DD\%hh-%mm-%ss` and tested that the output file used `%CCYY-%MM-%DD/Prefix %hh-%mm-%ss Suffix`.

 Set `Prefix` to `Prefix/` and tested that the output file used `%CCYY-%MM-%DD/Prefix/%hh-%mm-%ss Suffix`.

Set `Prefix` to `///Prefix/` and tested that the output file used `%CCYY-%MM-%DD/Prefix/%hh-%mm-%ss Suffix`.

 Set `Suffix` to `Suffix/` and tested that the output file used `%CCYY-%MM-%DD/Prefix/%hh-%mm-%ss Suffix_`. ( I'm not sure why a user would ever do this, but at least it won't crash)

 Set `Suffix` to `/Suffix/` and tested that the output file used `%CCYY-%MM-%DD/Prefix/%hh-%mm-%ss/Suffix_`.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

NikhilNarayana commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -832,8 +832,6 @@ bool SimpleOutput::StartStreaming(obs_service_t *service)
 
 static void remove_reserved_file_characters(string &s)
```
Is this function necessary? I couldn't find if regular recordings had some similar handling, but this function was only called by the replay start functions. Additionally it was the reason directories were being removed so I just deleted those cases but left the rest.

jp9000 replied:
I think it's used for replays because of the fact that replay suffixes (set in advanced) might contain invalid characters.

NikhilNarayana replied:
Gotcha, I can see that being in a issue, so I guess it is good that I left most of it intact.

UserNaem commented:
Have you verified that your change also respects the "Generate File Name without Space" setting in Settings > Output?

jp9000 commented:
This functions if the folders specified in the prefix exist, but will crash if the folders do not exist.  Likely you will have to ensure that the directories exist.

NikhilNarayana commented:
> This functions if the folders specified in the prefix exist, but will crash if the folders do not exist. Likely you will have to ensure that the directories exist.

@jp9000, I tested with Prefix set to `Prefix/` and the folder was created successfully because the new function I wrote gets called before `ensure_directory_exists` which handles creating directories if necessary. Suffix is a bit more tricky, users might want to set a directory in the suffix, but the suffix input box shouldn't allow them to end with a slash. I tried adding a validator with `.*[^\\,\/]$` as the regex but it isn't working. Any insight on that would be great. 

> Have you verified that your change also respects the "Generate File Name without Space" setting in Settings > Output?

Yes that works fine.

NikhilNarayana commented:
Fixed directories from prefix and suffix not being created in simple output mode and removed the ability for the suffix to break recordings if there is a slash at the end by replacing it with an underscore. I think a validator to prevent adding a slash at the end of the suffix would be a good idea, but my implementation isn't working (I tried mimicing the resolution validators).

jp9000 commented:
Unfortunately, your code still kept crashing.  I wrote my own fix in 002886bbac that was a bit more simple of a solution.  It now allows slashes in recording names, and it will automatically generate directories as needed for recording files and replay buffers.
